File: y2023/day24/main.go
package day24

import (
	"fmt"
	"strings"

	"adventofcode.com/internal/geom"
	"adventofcode.com/internal/math"
	"adventofcode.com/internal/utils"
)

type Equation []int

func (e Equation) Multiply(x int) {
	for i := range e {
		e[i] *= x
	}
}

func (e Equation) Sub(rhs Equation) {
	if len(e) != len(rhs) {
		panic(fmt.Sprint("Different equation lengths:", e, rhs))
	}
	for i := range e {
		e[i] -= rhs[i]
	}
}

func newEquation(a ...int) Equation {
	return a
}

func intersection2D(p1, p2 geom.PointInt, v1, v2 geom.VectorInt) (geom.PointFloat, float64, float64, bool) {
	// p1X + v1X * t1 = p2X + v2X * t2
	// p1Y * v1Y * t1 = p2Y + v2Y * t2
	// v1X * t1 + (-v2X) * t2 = p2X - p1X
	eqX := newEquation(v1.X, -v2.X, p2.X-p1.X)
	if eqX[0] < 0 {
		eqX.Multiply(-1)
	}
	eqY := newEquation(v1.Y, -v2.Y, p2.Y-p1.Y)
	if eqY[0] < 0 {
		eqY.Multiply(-1)
	}

	// a1 * t1 + a2 * t2 = a3
	// b1 * t1 + b2 * t2 = b3
	lcm := math.LCM(eqX[0], eqY[0])
	if lcm != 0 {
		k1 := lcm / eqX[0]
		k2 := lcm / eqY[0]
		eqX.Multiply(k1)
		eqY.Multiply(k2)
		eqX.Sub(eqY)
	}

	var t1, t2 float64
	if eqX[0] == 0 {
		if eqX[1] == 0 {
			return geom.PointFloat{}, 0, 0, false
		}
		t2 = float64(eqX[2]) / float64(eqX[1])
		t1 = (float64(eqY[2]) - t2*float64(eqY[1])) / float64(eqY[0])
	} else if eqY[0] == 0 {
		if eqY[1] == 0 {
			return geom.PointFloat{}, 0, 0, false
		}
		t2 = float64(eqY[2]) / float64(eqY[1])
		t1 = (float64(eqX[2]) - t2*float64(eqX[1])) / float64(eqX[0])
	} else {
		panic(fmt.Sprint("Error:", eqX, eqY))
	}

	x1 := float64(p1.X) + t1*float64(v1.X)
	y1 := float64(p1.Y) + t1*float64(v1.Y)
	return geom.PointFloat{X: x1, Y: y1}, t1, t2, true
}

func parseInput(input string) (pos []geom.PointInt, vel []geom.VectorInt) {
	for _, line := range utils.NonEmptyLines(input) {
		pStr, vStr, _ := strings.Cut(line, " @ ")
		p := utils.SplitNumbers(pStr, ",")
		pos = append(pos, geom.PointInt{X: p[0], Y: p[1], Z: p[2]})
		v := utils.SplitNumbers(vStr, ",")
		vel = append(vel, geom.VectorInt{X: v[0], Y: v[1], Z: v[2]})
	}
	return pos, vel
}

func solveV1(input string, mn, mx float64) int {
	pos, vel := parseInput(input)

	res := 0
	for i := range len(pos) {
		for j := range i {
			p, t1, t2, found := intersection2D(pos[i], pos[j], vel[i], vel[j])
			if found && t1 >= 0 && t2 >= 0 && mn <= p.X && p.X <= mx && mn <= p.Y && p.Y <= mx {
				res++
			}
		}
	}
	return res
}

func SolveV1(input string) int {
	return solveV1(input, 200000000000000, 400000000000000)
}

File: y2023/day24/main.py
import collections
import dataclasses
import random

import numpy as np
import scipy

from internal.py import task, task_io


@dataclasses.dataclass
class Point:
    x: int
    y: int
    z: int


@dataclasses.dataclass
class Vector:
    x: int
    y: int
    z: int


"""
https://www.reddit.com/r/adventofcode/comments/18q40he/2023_day_24_part_2_a_straightforward_nonsolver/
(dy'- dy) X + ( dx-dx') Y + (  y- y') DX + ( x'-  x) DY =  x' dy' -  y' dx' -   x  dy +   y  dx
(v2y-v1y) X + (v1x-v2x) Y + (p1y-p2y) DX + (p2x-p1x) DY = p2x v2y - p2y v2x - p1x v1y + p1y v1x
X, Y, DX, DY - initial position and velocity

Equation:
a0 X + a1 Y + a2 Z + a3 DX + a4 DY + a5 DZ = b
"""


def __make_equation(p1x, p1y, v1x, v1y, p2x, p2y, v2x, v2y) -> (list[int], int):
    a = [0] * 4
    a[0] = v2y - v1y
    a[1] = v1x - v2x
    a[2] = p1y - p2y
    a[3] = p2x - p1x
    b = p2x * v2y - p2y * v2x - p1x * v1y + p1y * v1x
    return a, b


def __make_equations(p1: Point, v1: Vector, p2: Point, v2: Vector) -> (list[list[int]], list[int]):
    a, b_xy = __make_equation(p1.x, p1.y, v1.x, v1.y, p2.x, p2.y, v2.x, v2.y)
    a_xy = [0] * 6
    a_xy[0], a_xy[1], a_xy[3], a_xy[4] = a

    a, b_xz = __make_equation(p1.x, p1.z, v1.x, v1.z, p2.x, p2.z, v2.x, v2.z)
    a_xz = [0] * 6
    a_xz[0], a_xz[2], a_xz[3], a_xz[5] = a
    return [a_xy, a_xz], [b_xy, b_xz]


def solve_v2(data: str) -> int:
    lines = task_io.non_empty_lines(data)

    points: list[Point] = []
    vecs: list[Vector] = []
    for line in lines:
        parts = line.split(" @ ")
        points.append(Point(*task_io.parse_numbers(parts[0], ", ")))
        vecs.append(Vector(*task_io.parse_numbers(parts[1], ", ")))

    results = collections.Counter()
    for _ in range(100):
        ind = random.sample(range(len(points)), k=4)
        aa, bb = [], []
        for i in range(1, 4):
            a, b = __make_equations(points[ind[0]], vecs[ind[0]], points[ind[i]], vecs[ind[i]])
            aa.extend(a)
            bb.extend(b)

        a = np.array(aa, dtype=np.longdouble)
        b = np.array(bb, dtype=np.longdouble)

        x = scipy.linalg.solve(a, b)
        res = int(np.round(x[0] + x[1] + x[2]))
        results[res] += 1

    # avoiding errors related to floating numbers accuracy
    return results.most_common(1)[0][0]


def solver(data_dir_path: str):
    task.solve(data_dir_path, solver_v2=solve_v2)


if __name__ == "__main__":
    solver("y2023/day24")
